### DISCUSSION TITLE
Polls architecture and implementation

### DISCUSSION THREAD

**Discussion by BartoszJarocki** [ParabolInc/parabol#5207] [Ideas]
Before I start writing code I'd love to get some feedback about things that are not clear to me. 
Polls epic: https://github.com/ParabolInc/parabol/issues/5147
Polls design: https://www.figma.com/file/D25gVH1t2alUiOdYNs5Owo/Multiple-content-input-types?node-id=518%3A19375 

## Design
I have a few questions about the design
1. When someone is creating a poll, are we supposed to show the status text 'X is creating a poll...' in a similar way we do this for tasks and comments? 
2. When I added an option and I want to remove it, how would I do that? I think we're missing a button to remove a single option when creating/editing a poll. 
3. Can we add reactions to a poll?

## Database schema 
When I thought about how the data model should look like I considered the following
1. We should now what's the status of the poll (ie. if it's active, ended, being edited)
2. User has only one vote in a poll 
3. Each poll is tied to the discussion

```sql
IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'PollStatus') THEN
  CREATE TYPE "PollStatus" AS ENUM (
    'active',
    'editing',
    'ended'
  );
END IF;

CREATE TABLE IF NOT EXISTS Poll {
  id VARCHAR(100) PRIMARY KEY
  created_by_id VARCHAR(100) REFERENCES User(id)
  discussion_id VARCHAR(100) REFERENCES Discussion(id)
  winning_option VARCHAR(100) REFERENCES PollOption(id)
  title VARCHAR(3000) 
  created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  status PollStatus
}
CREATE INDEX IF NOT EXISTS "idx_discussion_id" ON Poll(discussion_id);

CREATE TABLE IF NOT EXISTS PollOption {
  id VARCHAR(100) PRIMARY KEY
  poll_id VARCHAR(100) REFERENCES Poll(id)
  created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  total_votes int
  title VARCHAR(3000) 
 }
CREATE INDEX IF NOT EXISTS "idx_poll_id" ON PollOption(poll_id);
 
CREATE TABLE IF NOT EXISTS UserPollOptionSelection {
  id VARCHAR(100) PRIMARY KEY
  poll_id VARCHAR(100) REFERENCES Poll(id)
  user_id VARCHAR(100) REFERENCES User(id)
  poll_option_id VARCHAR(100) REFERENCES PollOption(id)
  created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
  UNIQUE (poll_id, user_id)
}
CREATE INDEX IF NOT EXISTS "idx_user_selection" ON UserPollOptionSelection(user_id, poll_id);
```

I'm not sure about the indexes and in general about the data model. I _think_ this covers all the use cases but I'd appreciate it if someone can take a look 🙏 .

BTW. I've also played with a data modeling tool: https://dbdiagram.io/d/60f97115b7279e4123365a07 Maybe it'd be useful to have our schema there.

## Backend
Similar to `Comment` and `Task`, `Poll` should implement the `Threadable` interface.
In the same manner, we should have a basic CRUD:
1. `AddPollMutation`
2. `EditPollMutation`
3. `DeletePollMutation`
4. `AddPollVoteMutation`
5. `RemovePollVoteMutation` - is the user able to remove the vote after submitting the vote?

In general, if I'm missing something or you think something could be improved please let me know. Any feedback is appreciated!

**Comment by enriquesanchez** [comment]
Thanks for putting this together, @BartoszJarocki 🙌


>When someone is creating a poll, are we supposed to show the status text 'X is creating a poll...' in a similar way we do this for tasks and comments?

Yes, it should work the same way it currently does for tasks and comments 👍

>When I added an option and I want to remove it, how would I do that? I think we're missing a button to remove a single option when creating/editing a poll.

If the user has not submitted the poll (by pressing the `Start` button), then all fields are still editable. I can click on any field (or <kbd>shift</kbd> + <kbd>TAB</kbd>) and change any of them.

<img width="525" alt="Screen Shot 2021-07-23 at 11 29 02 a m" src="https://user-images.githubusercontent.com/3276087/126813233-fca1a139-f61d-4177-b923-92f7eb941a10.png">


>Can we add reactions to a poll?

🤔 I did not think of this one. I wonder if the bring any value to a poll? Do you have any strong opinions? My gut tells me to go without reactions and add them later if we learn that we need them.

**Reply by BartoszJarocki** [reply]
@enriquesanchez thank you for checking it!

A few more questions, as I'm still not sure about how removing option fields works. 🙏 
> > When I added an option and I want to remove it, how would I do that? I think we're missing a button to remove a single option when creating/editing a poll.
> 
> If the user has not submitted the poll (by pressing the `Start` button), then all fields are still editable. I can click on any field (or shift + TAB) and change any of them.
> 
> <img alt="Screen Shot 2021-07-23 at 11 29 02 a m" width="525" src="https://user-images.githubusercontent.com/3276087/126813233-fca1a139-f61d-4177-b923-92f7eb941a10.png">
Given the screenshot you've attached, let's say I want to remove the last field (the empty one), how do I do that?
What happens if I click `+ Add another choice` if the previous one is still empty? 

> 🤔 I did not think of this one. I wonder if they bring any value to a poll? Do you have any strong opinions? My gut tells me to go without reactions and add them later if we learn that we need them.

No strong opinions here but initially, I'd go without reactions.

**Reply by enriquesanchez** [reply]
Ah, got it. Thanks for clarifying.

>Given the screenshot you've attached, let's say I want to remove the last field (the empty one), how do I do that?

You leave them empty. Only the first two are required for a poll to be created. The 3rd and 4th are optional, so if they are left empty, it's fine. How does that feel to you?

>What happens if I click + Add another choice if the previous one is still empty?

You get a 4th field. A good example of how this works is Twitter. If you create a poll, add the 1st and 2nd options, and then add a 3rd and 4th but leave them empty, you can still submit it. Only 2 are required. 

Now what if I add 4 fields, but only fill out the 1st and 3rd? Twitter won't let you create the poll because the 1st and 2nd fields are required.

Does that clarify your question?

**Comment by mattkrick** [comment]
schema looks good! couple thoughts:
- `UserPollOptionSelection` means only 1 user can vote on 1 category. when we use polls in slack today a user can vote on multiple fields & that seems like a pretty nice feature to have here, too. it'd be nice to have this business logic live outside the DB if possible so we could change it more easily.
- for the above, maybe we denormalize just a bit & put `voteUserIds[]` on `PollOption` instead of `total_votes` & we can get rid of the 3rd table. that way, if we want to force 1 vote per person, we just query all the options & make sure their ID doesn't exist in any of the `voteUserIds[]` otherwise, we just check the 1 `PollOption` that they're toggling their vote for
- let's use camelCase over snake_case. the added quotes in SQL queries are painful, but not as painful as mapping to camelCase on every query

For the mutations, I'd suggest:
- create poll
- remove poll
- add poll option
- remove poll option
- rename poll option
- vote for poll (isUnvote: boolean)

in the past i combined vote/unvote as a single mutation because aside from the query itself the logic is pretty similar (see https://github.com/ParabolInc/parabol/blob/b6c409332cd08bb48dbd505e10bcbe5ef1ea1cab/packages/server/graphql/mutations/voteForReflectionGroup.ts) but if it looks better as 2 mutations that's AOK

the big question in my mind is what should the other participants see while someone is creating a poll? they see "X is creating a poll" but do they see a wireframe? do they see options as they get added? if we stick with the start button, then we'll have to keep a bunch of local state until everything looks good & they hit "start". if we get rid of the start button, we can call create poll immediately & other team members see it being built in real time. this is similar to how tasks work today

 @enriquesanchez curious for your thoughts

**Reply by BartoszJarocki** [reply]
@mattkrick thank you! 

> UserPollOptionSelection means only 1 user can vote on 1 category. when we use polls in slack today a user can vote on multiple fields & that seems like a pretty nice feature to have here, too. it'd be nice to have this business logic live outside the DB if possible so we could change it more easily.

Makes sense. For some reason I had Twitter polls in mind and I assumed it'd work the same. I agree, it's best to keep these constraints outside of DB. 

> for the above, maybe we denormalize just a bit & put voteUserIds[] on PollOption instead of total_votes & we can get rid of the 3rd table. that way, if we want to force 1 vote per person, we just query all the options & make sure their ID doesn't exist in any of the voteUserIds[] otherwise, we just check the 1 PollOption that they're toggling their vote for

Right. I'll remove the table and add an array field to `PollOption` 

> let's use camelCase over snake_case. the added quotes in SQL queries are painful, but not as painful as mapping to camelCase on every query

Right, I'm used to snake_case in DB and TypeORM automatically converting that 😅  Will use camelCase.

Regarding the mutations, makes sense! I don't mind having one mutation for vote/unvote 👍 
EDIT: I think we're missing one for 'FinishPoll', if I remember correctly a creator should be able to finish the poll manually. Correct me if I'm wrong @enriquesanchez 

> the big question in my mind is what should the other participants see while someone is creating a poll? they see "X is creating a poll" but do they see a wireframe? do they see options as they get added? if we stick with the start button, then we'll have to keep a bunch of local state until everything looks good & they hit "start". if we get rid of the start button, we can call create poll immediately & other team members see it being built in real time. this is similar to how tasks work today

I personally think showing options as they get added is a bit creepy and I'd rather wait for the whole poll to be built. But it's just a personal preference.

**Reply by enriquesanchez** [reply]
>if I remember correctly a creator should be able to finish the poll manually. Correct me if I'm wrong @enriquesanchez

@BartoszJarocki Yes, a poll can be ended by the creator or the facilitator.

Re: the submit button and showing options as they get added:

Having a submit/create button was one of the topics that came up the most from the feedback I got. Because creating a poll requires filling out multiple fields it felt better to have the submit action. 

I agree that we should show the "X is creating a poll" message and the wireframe/empty card of the poll.

**Comment by BartoszJarocki** [comment]
One more question, if user deletes a poll, are we doing hard or soft?

**Reply by mattkrick** [reply]
always soft if possible
